File: jekyun-park/week5/programmers/92334.swift
//
//  main.swift
//  SwiftAlgorithms
//
//  Created by 박제균 on 2022/01/27.
//

import Foundation

func solution(_ id_list: [String], _ report: [String], _ k: Int) -> [Int] {

    var countDictionary: [String: Set<String>] = [:]
    var reportDictionary: [String: [String]] = [:]
    var answer: [Int] = []
    
    id_list.forEach {
        countDictionary[$0] = []
        reportDictionary[$0] = []
    }

    for content in report {
        let splitted = content.split(separator: " ")
        let reporter = String(splitted[0])
        let banned = String(splitted[1])
        countDictionary[banned]!.insert(reporter)
        if !reportDictionary[reporter]!.contains(banned) {
            reportDictionary[reporter]?.append(banned)
        }
    }

    id_list.forEach { answer.append((reportDictionary[$0]?.filter { countDictionary[$0]!.count >= k }.count)! ) }
    
    return answer
}

print(solution(["muzi", "frodo", "apeach", "neo"], ["muzi frodo", "apeach frodo", "frodo neo", "muzi neo", "apeach muzi"], 2))

//print(solution(["con", "ryan"], ["ryan con", "ryan con", "ryan con", "ryan con"], 3))

File: leehyowonzero/5week/92334.py
from collections import defaultdict

def solution(id_list, report, k):
    answer = defaultdict(int) # 정답 딕트 만들기
    for el in id_list:
        answer[el] = 0
        
    reported_id_list = defaultdict(set) # 신고당한 아이디 리스트 초기화 및 갱신, 중복 신고를 막기위해 key값은 set자료구조로 저장
    for case in report:
        fm, to = case.split(' ')
        reported_id_list[to] |= {fm}
        
    for reported_id  in reported_id_list: # 조건이 만족되면 정답 딕트에 적용
        if(len(reported_id_list[reported_id]) >= k):
            for el in reported_id_list[reported_id]:
                answer[el] += 1
    return list(answer.values())